src/mongocxx/test/v1/change_stream.cpp

```diff
@@ -0,0 +1,622 @@
+            CHECK(data.error_document_count == 1);
+            CHECK(stream.begin() == stream.end());
+        }
+    }
+}
+
+TEST_CASE("ownership", "[mongocxx][v1][change_stream]") {
+    struct test_data_type {
+        int destroyed_count = 0;
+    } source_data, target_data;
+
+    auto change_stream_destroy = libmongoc::change_stream_destroy.create_instance();
+    change_stream_destroy
+        ->interpose([&](mongoc_change_stream_t* stream) -> void {
+            auto const data = reinterpret_cast<test_data_type*>(stream);
+
+            if (data == &source_data || data == &target_data) {
+                data->destroyed_count += 1;
+            } else {
+                CHECK(static_cast<void const*>(&source_data));
+                CHECK(static_cast<void const*>(&target_data));
+                CHECK(static_cast<void const*>(data));
+                FAIL_CHECK("should not reach this point");
+            }
+        })
+        .forever();
+
+    auto source = change_stream::internal::make(reinterpret_cast<mongoc_change_stream_t*>(&source_data));
+    bsoncxx::v1::stdx::optional<change_stream> target =
+        change_stream::internal::make(reinterpret_cast<mongoc_change_stream_t*>(&target_data));
+
+    SECTION("move") {
+        auto move = std::move(source);
+
+        CHECK(source_data.destroyed_count == 0);
+        CHECK(target_data.destroyed_count == 0);
+
+        target = std::move(move);
+
+        CHECK(source_data.destroyed_count == 0);
+        CHECK(target_data.destroyed_count == 1);
+
+        target.reset();
+
+        CHECK(source_data.destroyed_count == 1);
+        CHECK(target_data.destroyed_count == 1);
+    }
+}
+
+TEST_CASE("end", "[mongocxx][v1][change_stream]") {
+    struct test_data_type {
+        int next_count = 0;
+    } data;
+
+    auto change_stream_destroy = libmongoc::change_stream_destroy.create_instance();
+    change_stream_destroy
+        ->interpose([&](mongoc_change_stream_t* p) -> void { CHECK(reinterpret_cast<test_data_type*>(p) == &data); })
+        .forever();
+
+    auto change_stream_next = libmongoc::change_stream_next.create_instance();
+    change_stream_next
+        ->interpose([&](mongoc_change_stream_t* stream, bson_t const** bson) -> bool {
+            (void)stream;
+            (void)bson;
+            FAIL("should not reach this point");
+            return false;
+        })
+        .forever();
+
+    auto const s1 = change_stream::internal::make(reinterpret_cast<mongoc_change_stream_t*>(&data));
+    auto const s2 = change_stream::internal::make(reinterpret_cast<mongoc_change_stream_t*>(&data));
+
+    CHECK(s1.end() == s1.end());
+    CHECK(s1.end() == s2.end());
+    CHECK(s1.end() == change_stream::iterator{});
+    CHECK(change_stream::iterator{} == change_stream::iterator{});
+
+    CHECK((++s1.end()) == s1.end());
+    CHECK((s1.end()++) == s1.end());
+
+    CHECK((*s1.end()).empty()); // OK.
+    // CHECK(s1.end()->empty()); // Undefined behavior.
```

Comment:
No requested change. TIL dereferencing the end iterator for `v_noabi` returns the current document of the underlying change stream:
```c++
auto iter = cs.begin();
REQUIRE((*iter)["fullDocument"]["a"].get_int32().value == 1);

auto doc = *cs.end();
REQUIRE(doc["fullDocument"]["a"].get_int32().value == 1);
```

I would rather not preserve that behavior, but may be worth noting if changes between the `v_noabi` and `v1` are being documented in release notes.